### ISSUE TITLE
Convert test infa to something more reliable and new 

### ISSUE THREAD

**Issue by vanesyan** [tj/commander.js#649]
commander.js has pretty old test infra based on sinon and should, we can convert it to something more popular. With convertion we may archive the following goals: keep it up to date to today's requirements, to make it easier to make contributions and convert it to es6 (that plan, I has in my mind), and, finally, to make it cross-platform. The current test infra can only be run on *nix systems, but no way to run test on env. that has no bash (or similar shells) and make, like windows.

**Comment by jseminck** [comment]
FYI I am trying to run the tests and do not even manage to get it working, out of the box:

```
commander.js $ npm run test

> commander@2.10.0 test /Users/joachimseminck/Projects/commander.js
> make test


   test.arguments.js ✖

  error: unknown option `--setup_mode'

make: *** [test] Error 1

npm ERR! Darwin 15.5.0
npm ERR! argv "/Users/joachimseminck/.nvm/versions/node/v7.1.0/bin/node" "/Users/joachimseminck/.nvm/versions/node/v7.1.0/bin/npm" "run" "test"
npm ERR! node v7.1.0
npm ERR! npm  v3.10.9
npm ERR! code ELIFECYCLE
npm ERR! commander@2.10.0 test: `make test`
npm ERR! Exit status 2
```

**Comment by CheshireSwift** [comment]
Hi, I've started working on this. I'm looking to move to mocha as the test framework, just replacing the Makefile + run.

It should:
* Make contributions easier: tests are descriptively named.
* Make it cross platform: mocha itself is pure Node, and I'll be structuring it such that the test command doesn't use globs or other POSIX magic (to keep Windows happy).

In this first pass I won't actually be changing any of the test logic, other than sandboxing program config (i.e. using `new Command()` for each test), so there's no ES6 rewrites in it yet, though I'll be using lambdas to define the tests.

Shout if there's any issues with any of this! 😃

**Comment by mojavelinux** [comment]
Although I like mocha, I think [ava](https://github.com/avajs/ava) would be the best fit for this project. It's extremely simple and lightweight, and it can continue the practice of running the test files in separate node processes.

To help move this discussion forward, I started PR #755 with just two of the test files migrated. This should hopefully provide enough context to help the group come to a decision about whether or not this is the right choice.

> using new Command() for each test

I totally agree we should be using `new Command()` for the test instead of using the global instance. That will make testing multiple scenarios in the same file feasible.

> so there's no ES6 rewrites in it yet, though I'll be using lambdas to define the tests.

Given that 4 is the minimum node version supported, there's no reason not to use lambdas at least.

**Comment by JohnRGee** [comment]
Adding some notes for reference.

Mocha and AVA are mentioned in this issue, and @mojavelinux has prototyped some tests with AVA in #755. (Thanks! Nice demo without wasting effort if go with something else.)

Jest is mentioned in #661. 

I use Jest for testing my own CLI project including commander, and have not used AVA or Mocha.

For interest, I checked weekly downloads from npmjs.com:
- AVA: 149 k weekly downloads
- Mocha: 2.5 M weekly downloads
- Jest: 4.4 M weekly downloads (and growing the fastest)

https://stackshare.io/stackups/ava-vs-jest

(Side note: tests can run on Windows already with a little work! #916)

**Comment by vlikin** [comment]
I am for mocha because it is almost every where on projects where I work. Jest is from React world.

**Comment by shadowspawn** [comment]
@mojavelinux converted a couple of the test files to ava in #755 to show what it looks like, which I thought was a useful way of seeing real code for a modest time investment. 

If someone is interested in showing what it would look like using `jest`, you could do a draft Pull Request with the same two files and a `jest` setup.

**Comment by shadowspawn** [comment]
Current contenders are AVA (#755 ) and Jest (#1005).

**Comment by shadowspawn** [comment]
Started move to Jest in #1035.

**Comment by shadowspawn** [comment]
Moved to Jest on the develop branch for the next release (#1035).

**Comment by shadowspawn** [comment]
v4.0.0-0 prerelease published: #1067

**Comment by shadowspawn** [comment]
v4.0.0 has been released.
